### ISSUE TITLE
SIGSEGV_HEAPCORRUPT_class2_cs_d!DomainBoundILStubClass.(ValX1`1<System.String>)

### ISSUE THREAD

**Issue by schaabs** [dotnet/coreclr#3728]
The notes in this bug refer to the dump [projectk-23910-00-amd64ret_004C](https://dotnetjobresults.blob.core.windows.net/projectk-23910-00-stress/d2ff25ad-766b-47a0-b18e-877d4fba3e73/stress/generated/06_02hr/projectk-23910-00-amd64ret_004C/projectk-23910-00-amd64ret_004C.csproj/projectk-23910-00-amd64ret_004C.zip?sv=2015-04-05&sr=c&sig=frXtec4JmbKXiwfVlQolyPLL07CGNB6LrdjCWWIJ%2F%2Fg%3D&se=2017-03-14T20:17:33Z&sp=rw)

**This issue is a segfault in the mark phase of the gc, caused by a corrupted root in DomainBoundILStubClass.(ValX1`1<System.String>).**
    
    FAILURE_HASH:
    SIGSEGV_HEAPCORRUPT_class2_cs_d!DomainBoundILStubClass.(ValX1`1<System.String>)

    STOP_REASON:
    SIGSEGV

    FAULT_SYMBOL:
    libcoreclr.so!WKS::gc_heap::mark_object_simple

    FAULT_STACK:
    libcoreclr.so!WKS::gc_heap::mark_object_simple
    libcoreclr.so!WKS::gc_heap::mark_object_simple
    libcoreclr.so!WKS::gc_heap::mark_object_simple
    libcoreclr.so!WKS::gc_heap::mark_object_simple
    libcoreclr.so!WKS::GCHeap::Promote
    libcoreclr.so!GcInfoDecoder::EnumerateLiveSlots
    libcoreclr.so!EECodeManager::EnumGcRefs
    libcoreclr.so!GcStackCrawlCallBack
    libcoreclr.so!Thread::MakeStackwalkerCallback
    libcoreclr.so!Thread::StackWalkFramesEx
    libcoreclr.so!Thread::StackWalkFrames
    libcoreclr.so!GCToEEInterface::GcScanRoots
    libcoreclr.so!GCToEEInterface::GcScanRoots
    libcoreclr.so!WKS::gc_heap::mark_phase
    libcoreclr.so!WKS::gc_heap::gc1
    libcoreclr.so!WKS::gc_heap::garbage_collect
    libcoreclr.so!WKS::GCHeap::GarbageCollectGeneration
    libcoreclr.so!WKS::gc_heap::try_allocate_more_space
    libcoreclr.so!WKS::GCHeap::Alloc
    ...

    CORRUPT_ROOT_FRAME_PC:
    0x7fe687feadcf

    CORRUPT_ROOT_THREAD:
    thread #11: tid = 14625, 0x00007fe7026ed414 libpthread.so.0`__pthread_cond_wait + 196 at pthread_cond_wait.S:185, stop reason = signal SIGSEGV

    CORRUPT_ROOT_FRAME:
    class2_cs_d!DomainBoundILStubClass.(ValX1`1<System.String>)

**Looking at the object we are marking it seems like it has a corrupted MethodTable**
     
    (lldb) up
    frame #3: 0x00007fe70142cc9f libcoreclr.so`WKS::gc_heap::mark_object_simple(po=<unavailable>) + 575 at gc.cpp:17950 [opt]
    (lldb) fr v -D 1
    (uint8_t **) po = <variable not available>

    (uint8_t *) o = 0x00007fe6606534c8 "\xb9\xb7\r`
    (size_t) s = 23296014613542
    ...

    --Dump the object to get it's method table
    (lldb) memory read --size 8 --format x --count 4 0x00007fe6606534c8
    0x7fe6606534c8: 0x00007fe6600db7b9 0x00007fe6600db7f8
    0x7fe6606534d8: 0x00007fe6600db838 0x00007fe6600db878
    
    --Dumping the method table it looks corrupt like it's had a partial address written to it
    (lldb) memory read --size 8 --format x --count 8 0x00007fe6600db7b9
    0x7fe6600db7b9: 0x1100007fe687dc39 0x6d006f0053000000
    0x7fe6600db7c9: 0x7300650054006500 0x7200740053007400
    0x7fe6600db7d9: 0x310067006e006900 0x0000000034003400
    0x7fe6600db7e9: 0x0000000000000000 0x7800000000000000

**Looking at the thread we're walking, it seems to be paused in an unconventional way through a native callout from a DomainBoundILStubClass**

    (lldb) thread select 11
    * thread #11: tid = 14625, 0x00007fe7026ed414 libpthread.so.0`__pthread_cond_wait + 196 at pthread_cond_wait.S:185, stop reason = signal SIGSEGV
        frame #0: 0x00007fe7026ed414 libpthread.so.0`__pthread_cond_wait + 196 at pthread_cond_wait.S:185
    (lldb) btm
    0       0x7fe7026ed414 libpthread.so.0!__pthread_cond_wait
    1       0x7fe70170a2f2 libcoreclr.so!CorUnix::CPalSynchronizationManager::ThreadNativeWait(CorUnix::_ThreadNativeWaitData*, unsigned int, CorUnix::ThreadWakeupReason*, unsigned int*)
    2       0x7fe701709fb0 libcoreclr.so!CorUnix::CPalSynchronizationManager::BlockThread(CorUnix::CPalThread*, unsigned int, bool, bool, CorUnix::ThreadWakeupReason*, unsigned int*)
    3       0x7fe70170ea91 libcoreclr.so!CorUnix::InternalWaitForMultipleObjectsEx(CorUnix::CPalThread*, unsigned int, void* const*, int, unsigned int, int)
    4       0x7fe70170ebb2 libcoreclr.so!WaitForSingleObjectEx
    5       0x7fe701327a21 libcoreclr.so!CLREventBase::WaitEx(unsigned int, WaitMode, PendingSync*)
    6       0x7fe701327a1c libcoreclr.so!CLREventBase::WaitEx(unsigned int, WaitMode, PendingSync*)
    7       0x7fe701327a13 libcoreclr.so!CLREventBase::WaitEx(unsigned int, WaitMode, PendingSync*)
    8       0x7fe7013279b0 libcoreclr.so!CLREventBase::WaitEx(unsigned int, WaitMode, PendingSync*)
    9       0x7fe70132c1b3 libcoreclr.so!Thread::RareDisablePreemptiveGC()
    10      0x7fe70132f6a5 libcoreclr.so!HandleGCSuspensionForInterruptedThread(_CONTEXT*)
    11      0x7fe70132f688 libcoreclr.so!HandleGCSuspensionForInterruptedThread(_CONTEXT*)
    12      0x7fe70132f65d libcoreclr.so!HandleGCSuspensionForInterruptedThread(_CONTEXT*)
    13      0x7fe7016def2d libcoreclr.so!inject_activation_handler(int, siginfo_t*, void*)
    14      0x7fe7026f1340 libpthread.so.0!???
    15      0x7fe687feadcf class2_cs_d!DomainBoundILStubClass.(ValX1`1<System.String>)
    16      0x7fe687fe8bf1 mscorlib.ni!System.Collections.Generic.GenericArraySortHelper`1[[ValX1`1[[System.__Canon, mscorlib]], SortedListTests]].BinarySearch(ValX1`1<System.__Canon>[], Int32, Int32, ValX1`1<System.__Canon>)
    17      0x7fe687fe7b90 mscorlib.ni!System.Collections.Generic.GenericArraySortHelper`1[[ValX1`1[[System.__Canon, mscorlib]], SortedListTests]].BinarySearch(ValX1`1<System.__Canon>[], Int32, Int32, ValX1`1<System.__Canon>, System.Collections.Generic.IComparer`1<ValX1`1<System.__Canon>>)
    18      0x7fe687fe728d mscorlib.ni!System.Array.BinarySearch[[ValX1`1[[System.__Canon, mscorlib]], SortedListTests]](ValX1`1<System.__Canon>[], Int32, Int32, ValX1`1<System.__Canon>, System.Collections.Generic.IComparer`1<ValX1`1<System.__Canon>>)
    19      0x7fe687fe6b40 System.Collections!System.Collections.Generic.SortedList`2[[ValX1`1[[System.__Canon, mscorlib]], SortedListTests],[System.__Canon, mscorlib]].Add(ValX1`1<System.__Canon>, System.__Canon)
    20      0x7fe687fe48dc SortedListTests!SortedListIdxValue.Driver`4[[ValX1`1[[System.__Canon, mscorlib]], SortedListTests],[System.__Canon, mscorlib],[System.__Canon, mscorlib],[System.Int32, mscorlib]].BasicIndexOfValue(ValX1`1<System.__Canon>[], System.__Canon[])
    21      0x7fe687fbfeab SortedListTests!SortedListIdxValue.IndexOfValue.IndexOfValueMain()
    22      0x7fe687fbd8be projectk-23910-00-amd64ret_004C!stress.generated.UnitTests.UTE()
    23      0x7fe687fbcf58 stress.execution!stress.execution.UnitTest.Execute()
    
**OTHER AVAILABLE DUMPS**
[projectk-23910-00-amd64chk_006D](https://dotnetjobresults.blob.core.windows.net/projectk-23910-00-stress/f2e8b09f-d887-43f7-baef-f6f1810fc6a0/stress/generated/07_02hr/projectk-23910-00-amd64chk_006D/projectk-23910-00-amd64chk_006D.csproj/projectk-23910-00-amd64chk_006D.zip?sv=2015-04-05&sr=c&sig=frXtec4JmbKXiwfVlQolyPLL07CGNB6LrdjCWWIJ%2F%2Fg%3D&se=2017-03-14T20:17:33Z&sp=rw)

**Comment by janvorli** [comment]
I am investigating. I can reliably repro it locally and it is a GC hole, the corrupted MethodTable is just a coincidence of the original object location and the data that were written to that memory after the object was relocated. The reference in the stack slot in the DomainBoundILStubClass.(ValX1`1<System.String>) was not updated for some reason when the actual object moved.
In all the repros, it happened when the thread was redirected while executing some DomainBoundILStubClass function (not always the same one).

**Comment by janvorli** [comment]
So, it turns out it is a memory corruption of a very interesting kind. Something corrupts the object address by setting bit zero to 1. The DomainBoundILStubClass method gets a reference to an element inside of an array. And that element contains a reference that is corrupted as I've said. 
@schaabs your analysis was correct upto this point:
```
--Dump the object to get it's method table
(lldb) memory read --size 8 --format x --count 4 0x00007fe6606534c8
0x7fe6606534c8: 0x00007fe6600db7b9 0x00007fe6600db7f8
0x7fe6606534d8: 0x00007fe6600db838 0x00007fe6600db878
```
But then you've tried to dump memory at the address 0x00007fe6600db7b**9** assuming it is a method table pointer. But that address itself is the corrupted one. All objects start at addresses aligned to 8 bytes. So the proper address of the object is 0x00007fe6600db7b**8** and there you can see a valid object.
I have reproed the problem several times and the corruption is always in the bit zero of an element in an array, but the element index is different each time and the array itself is also different.

**Comment by janvorli** [comment]
When I've said that the array is always different, I meant a different array in a different method.

**Comment by janvorli** [comment]
Fixed by #3812
